### PR TITLE
netatalk: Update to 4.0.4

### PR DESCRIPTION
**Pull request by APCCV**
Maintainer: Antonio Pastor / @APCCV
Compile tested: ipq806x (23.05.3, snapshot)
Run tested: ipq806x - C2600: start server, connect from MacOS, read/write files to home share, create/update TimeMachine backups

Description:
No changes to package other than using latest available upstream code base. Starting Netatalk 4.x build uses meson instead of autotools.

### CODE REVIEW

File: net/netatalk/Makefile
Comment (BKPepe) [inline net/netatalk/Makefile:91]: Why added indentation here? Thats not correct.

**Review by BKPepe — CHANGES_REQUESTED** [review]
So many unrelevant changes, which are not described anywhere.

1. Dropping LICENSE
2. Adding subpackage
3. Adding indentation, which then makes it completely useless.
etc

File: net/netatalk/Makefile
Comment (APCCV) [reply net/netatalk/Makefile:91]: I formatted everything thinking all `define Package` had tab-nested comments.
I see now this one in particular is different, no indent.
I see half of the sections have only two-space indent.
Will update overall.

**Comment by APCCV** [comment]
Updated indents. Comments on your change request.

The so many irrelevant changes are my attempt to clean up this old Makefile. Looked at other Makefiles in different libraries to match as best as it made sense to me.

Licence not dropped. Just moved to a more common location.

Added a download. Not a sub-package.
Upstream dropped a file (Unicode DB) that is used in build and made it a pre-req. Did not find any other package that provides this file. It is only used during build. A 2.2MB, it is quite large so I considered it to be a better solution to download it on the fly than to store it in GitHub. Having it in Makefile makes it slightly easier to update if a new version of the Unicode DB were to be released.
Added a comment to Makefile for it.
If you think a different approach is needed let me know which (and if you don't mind, why).

**Comment by APCCV** [comment]
As the commit message states, upstream moved from autotools to meson. That is the bulk of the relevant changes.

**Comment by APCCV** [comment]
4.0.3 just released. Updated version and download hash.

**Comment by Neustradamus** [comment]
@APCCV: Very fast update, thanks :)

File: net/netatalk/Makefile
Comment (BKPepe) [inline net/netatalk/Makefile:33]: Why do we need these two variables with meson? :)

```suggestion
```

File: net/netatalk/Makefile
Comment (BKPepe) [inline net/netatalk/Makefile:40]: In my case, this Title is too long and thus it is not disaplayed in make menuconfig.

File: net/netatalk/Makefile
Comment (APCCV) [reply net/netatalk/Makefile:33]: Yes. Will do.

File: net/netatalk/Makefile
Comment (APCCV) [reply net/netatalk/Makefile:34]: Agree. As above.

File: net/netatalk/Makefile
Comment (APCCV) [reply net/netatalk/Makefile:40]: Will revise. I got too creative. It just said "netatalk" before. Can make it half of that no problem.

**Comment by APCCV** [comment]
>     * How would you even version downloaded unicode?

It can be versioned. I can add an additional variable for version. Should have done that, like this (note `UC_DATA_VERS`):
```
# Unicode DB file required for compilation
UC_DATA_FILE:=UnicodeData.txt
UC_DATA_VERS:=16.0.0
UC_DATA_URL:=https://www.unicode.org/Public/$(UC_DATA_VERS)/ucd/
UC_DATA_HASH:=ff58e5823bd095166564a006e47d111130813dcf8bf234ef79fa51a870edb48f
```

> Do it in separate Makefile, please. = Hopefully, in the future, more packages will use it. Its logical to have it separately, because I am not quite sure why do you need to download it as subpackage if multiple GNU/Linux distributions such as Debian have that file in its own package. E.g. https://packages.debian.org/sid/unicode-data
> 
This is where I would rather debate the point.
Yes, many distributions have this file installed already. My first PR, which I cancelled, failed tests because of it. On my computer the package builds because the build finds the file on my distribution, outside the build "sandbox".
I think that is not correct. I think a package build should be self sufficient without requiring files from the host OS.
And again, the OpenWrt build failed because the containers don't have the file as they don't have the `https://packages.debian.org/sid/unicode-data` package installed (or equivalent for Alpine or whatever they use). Should it be installed there? Can it? For this (small) thing? How does that get done? Do an apt install during `Prepare`?

You have a point with reuse, but I doubt many other packages would use this. The issue w/netatalk is how the upstream used to have this file as part of the tarball, but now dropped it with no mechanism to download it as part of the build. How does OpenWrt packages normally deal with this? Can you think of an example so I take a look?

I am not to sure what you mean with "a separate Makefile." That's on me, I'm far from an expert with OpenWrt packages.
- Do you mean creating a separate "UnicodeData" package? That will not work. We don't want this raw file on the routers. We need the compiled version, and the build process does that. New package that downloads this raw file, and make it just a build dependency?
- Do you mean as a separate build process that is called before the package build to download the file and place it somewhere on the OS? I have no clue how to do that, but I can search and learn if that is the case. It would be awesome if you could point me to an existing package that has two Makefiles. I'll search for one in the meantime.
- Something else?

Thanks for your time and change requests to clean this up.

File: net/netatalk/Makefile
Comment (APCCV) [reply net/netatalk/Makefile:33]: Well, `PKG_BUILD_PARALLEL:=1` should stay. If not `include\package.mk` will just make singlethreaded.

Comment (APCCV) [reply net/netatalk/Makefile]: Tested with & without `PKG_BUILD_PARALLEL` and saw no difference. This package is so small it wouldn't matter anyways. Removed.

File: net/netatalk/Makefile
Comment (APCCV) [reply net/netatalk/Makefile]: Revised. Tested. Fits OK in menuconfig.

File: net/netatalk/Makefile
Comment (APCCV) [reply net/netatalk/Makefile]: Done.

**Comment by APCCV** [comment]
Alright. Summary:
- two commits, as requested, one with change to meson, one with cosmetic stuff
- eliminated download of additional file so no additional makefile or package

Please take a look.

File: net/netatalk/Makefile
Comment (BKPepe) [reply net/netatalk/Makefile]: No, no. Thats not correct. Your package is using meson and ninja. While checking CI log, you can see it there:
> 2024-10-29T02:57:34.4488563Z MAKEFLAGS="--jobserver-auth=3,4" /builder/staging_dir/host/bin/ninja   -C /builder/build_dir/target-arm_cortex-a9+vfpv3-d16_musl_eabi/netatalk-4.0.3/openwrt-build 
2024-10-29T02:57:34.4546074Z ninja: Entering directory `/builder/build_dir/target-arm_cortex-a9+vfpv3-d16_musl_eabi/netatalk-4.0.3/openwrt-build'

Also, if you look at this commit: https://github.com/openwrt/openwrt/commit/97258f53634d7237a2962aec3387f011047ce83b, if the package uses ninja, then it is using PKG_BUILD_PARALLEL. 

Also, in the CI/CI log, you can see that there is also cmake for netatalk.

File: net/netatalk/Makefile
Comment (BKPepe) [reply net/netatalk/Makefile]: Tabs - inherited from make, where commands needs to be indented with tabs.
Conffiles should not have any indentation, otherwise it does not work. You can try it yourself on the router or also by checking the .ipk with mc utility.

**Comment by BKPepe** [comment]
> * Do you mean creating a separate "UnicodeData" package? That will not work. We don't want this raw file on the routers. We need the compiled version, and the build process does that. New package that downloads this raw file, and make it just a build dependency?

Yes, I mean to create that package and mark it as `BUILDONLY:=1` e.g. how it is done in **nacl Makefile** in this repo.

Since you found out the file in the build system, then great! :) I havent checked it myself so far. I would do that if I there is need to create a new Makefile for the unciode.

File: net/netatalk/Makefile
Comment (BKPepe) [inline net/netatalk/Makefile:60]: I think, we can do this even better. Not tested, so far.

	-Dwith-unicode-data-path="$(STAGING_DIR_HOST)/share/gnulib/tests/uniname/" \

**Review by BKPepe — APPROVED** [review]
I really appreciate that you weren’t afraid to tackle this, found a solution on your own, and even tested it to see if it actually works. This is something you don’t see often these days, as people tend to copy things from the Internet without trying them out. It’s clear that you have patience and determination. This approach is rare and truly excellent. This will greatly help you in your journey of contributing to OpenWrt. Thank you for your work

**Comment by APCCV** [comment]
> Thank you for your work

Thank you for your patience.

Since you haven't merged the PR yet, I'm committing an additional change to enable the old AppleTalk protocol. I held it back because it needed patching or waiting for the next release.
AppleTalk is really the key enhancement for v4. It doesn't add to much in space.
I'd be surprised if more than 2 people end up using it. If anyone does, they are absolutely going to love that their OpenWrt router now networks all Macs.

**Comment by APCCV** [comment]
Alright. Really done with 4.0.3. Won't touch it anymore.
Please merge if you have no further concerns.

File: net/netatalk/Makefile
Comment (APCCV) [reply net/netatalk/Makefile]: Addressed in third commit that enables AppleTalk features.

**Comment by APCCV** [comment]
> Won't touch it anymore

Famous last words.
Bumped to 4.0.4 as it is now available. Removed patch I had in here before as those changes were adopted upstream.

Please merge if you are OK with PR as it stands now.

**Comment by APCCV** [comment]
` make[1]: *** [package/Makefile:179: package/kernel/linux/compile] Error 1`

Aw, come on...

**Comment by APCCV** [comment]
:(

**Review by BKPepe — APPROVED** [review]

**Comment by BKPepe** [comment]
CI/CD failures are not relevant to this PR. Merging, thank you.

**Comment by Neustradamus** [comment]
Good job, thanks to @APCCV, and @BKPepe for merging!

Can be backported to 23.05.x and 24.10.x?

**Comment by APCCV** [comment]
> Can be backported to 23.05.x and 24.10.x?

I have no problem with it. 
This package is all userspace... so I think you can take the snapshot package and install in 23.05... haven't tested that yet, but will give that a try. 
What I really want is for it to be in the imagebuilder (makes life simpler).

For next release I'm planning on splitting in small/large packages. Anyone having the resources (and an x86) to run the build with everything will probably be better off with the docker image. For smaller devices a -small package with only afpd will be more useful.

So how do we know if I can backport it? Just open the PR?

**Comment by APCCV** [comment]
Opened PR against 24.10 branch.
https://github.com/openwrt/packages/pull/25246

**Comment by hnyman** [comment]
Buildbot finds new dependencies and 4.0.4 does not compile...
New detected dependencies/features should be either disabled in the package configuration, or the correct dependencies added.

E.g.
https://downloads.openwrt.org/snapshots/faillogs/aarch64_cortex-a72/packages/netatalk/compile.txt

```
Package netatalk is missing dependencies for the following libraries:
libgio-2.0.so.0
libglib-2.0.so.0
libgobject-2.0.so.0
make[3]: *** [Makefile:122: /builder/shared-workdir/build/sdk/bin/packages/aarch64_cortex-a72/packages/netatalk_4.0.4-r1_aarch64_cortex-a72.ipk] Error 1
```

Backporting to 24.10 would make sense after it first compiles ok in master...
(but apparently this has already been backported to 24.10. Compilation will fail there too.)

**Comment by APCCV** [comment]
This was building properly. 
Apologies.
I will rebuild from scratch and have a PR to address as soon as possible.

**Comment by hnyman** [comment]
> This was building properly.
> Apologies.

No trouble.
It is just that as the buildbot builds all packages/libraries, there are much more libraries available than in your own build system (or in the PR CI testing). So when some additional features have been introduced in a package's new version, it may happen that new dependencies are detected only in buildbot.
This happens every now and then.

It pays off to monitor the buildbot faillogs after a PR gets merged.

> I will rebuild from scratch and have a PR to address as soon as possible.

You just need to either add those detected dependencies (possibly bloating the needed amount of libraries) ot adjust the build configure step of the package to leave those detected libraries unused.

**Comment by hnyman** [comment]
Yeah, first to main/master

**Comment by BKPepe** [comment]
Yes, PR on master first.

Dne po 4. 11. 2024 19:06 uživatel AP ***@***.***> napsal:

**Comment by APCCV** [comment]
Submitted https://github.com/openwrt/packages/pull/25251

**Comment by APCCV** [comment]
Where can I see the buildbot log for this? The one asking for those additional libraries?
I assume somewhere here https://buildbot.openwrt.org/master/packages/#/grid but so far I haven't found it.
Any pointers are welcome.

**Comment by hnyman** [comment]
> Where can I see the buildbot log for this? The one asking for those additional libraries?

How about the link I gave above?
to the "faillogs".
All architectures are here:
https://downloads.openwrt.org/snapshots/faillogs/

E.g. these two builds below are from today. But based on timestamps, they are from time before I merged your patch attempt.  
https://downloads.openwrt.org/snapshots/faillogs/aarch64_cortex-a76/packages/netatalk/compile.txt
https://downloads.openwrt.org/snapshots/faillogs/mipsel_24kc/packages/netatalk/compile.txt

You may need to wait for tomorrow, as today the whole buildbot infra was upgraded , and the phase2 packages buildbot is not really visible yet. 
(Or at least I haven't been able to find it...  @ynezz might know more: https://lists.openwrt.org/pipermail/openwrt-devel/2024-November/043358.html )

**Comment by APCCV** [comment]
> How about the link I gave above?

Thank you. Noticed that after I had posted my comment after a long trip through the build bot site. I know where to look now. Should've edited the comment.

> E.g. these two builds below are from today. 

I don't think so, based on timestamps. I'm on UTC-5 (and maybe that doesn't matter) and build logs for snapshot packages stopped at Tue Nov 5 09:40:44 2024 (for yesterday's batch) to start again at Tue Nov 5 23:26:46.
There's no failed log for arm_arm926ej-s, for example, which ended at Tue Nov 5 23:58:28 it seems.
And package is there: https://downloads.openwrt.org/snapshots//packages/arm_arm926ej-s/packages/netatalk_4.0.4-r1_arm_arm926ej-s.ipk with date Tue Nov 5 21:58:00 2024.

I waited for builds to start succeeding before I created the PR. I don't like breaking builds (and I know it's not of much consequence in this case).

Anyways, no rush. PR is there and should fix my mess in 24.10. Please merge when you think is prudent to.

**Comment by hnyman** [comment]
Thanks, I merged the backport.

**Comment by Neustradamus** [comment]
Thanks to @APCCV, @BKPepe for merging and @hnyman for review!

Original PRs in master:
- https://github.com/openwrt/packages/pull/25200 (this 4.0.4 PR)
- https://github.com/openwrt/packages/pull/25251 (the 4.0.4 PR fix)

24.10 backport:
- https://github.com/openwrt/packages/pull/25246 (of this 4.0.4 PR)
- https://github.com/openwrt/packages/pull/25260 (of the 4.0.4 PR fix)

**Comment by Neustradamus** [comment]
@APCCV, @BKPepe, @hnyman: It is possible to backport for 23.05.x too?

Thanks in advance.

**Comment by APCCV** [comment]
> It is possible to backport for 23.05.x too?

Backporting it would require 6 commits from when I took over the package in June. I can create a PR with that.
---OR--- 
I can port it by creating a PR with a new commit with the current Makefile. 

Is either acceptable?
I'll create the backport PR next Monday unless I hear otherwise.

**Comment by Neustradamus** [comment]
@APCCV: 2 new versions after Netatalk 4.0.4 -> 4.0.5 and 4.0.6 :)

Can you look?

Links:
- https://github.com/Netatalk/netatalk/releases/tag/netatalk-4-0-4
- https://github.com/Netatalk/netatalk/releases/tag/netatalk-4-0-5
- https://github.com/Netatalk/netatalk/releases/tag/netatalk-4-0-6

**Comment by APCCV** [comment]
@Neustradamus saw those. I get an email on new releases. rdmark also reached out about 4.0.5 for reasons below.
> * https://github.com/Netatalk/netatalk/releases/tag/netatalk-4-0-4
This is what we have in progress.
> * https://github.com/Netatalk/netatalk/releases/tag/netatalk-4-0-5
Ah, the irony. Honestly it was painful to deal with the Unicode txt file, first to figure how the … to make it work, then to make it OpenWrt compliant with an extra package (almost got there), then to make it work with a file from host (duh - had I known). In the meantime I exchanged notes with rdmark. Turns out OpenWrt wasn’t the only packaging that was going through this particular circle of hell, and now the file is optional.
Beyond that the fixes are mostly irrelevant to OpenWrt.
> * https://github.com/Netatalk/netatalk/releases/tag/netatalk-4-0-6
Also has little to offer over 4.0.4 unless using a System 7 client over TCP (which I can’t test).

I’ll keep an eye on releases for something major (in substance, not in version numbering) and will consider updating then. In the meantime I’d rather see 4.0.4 backported to 23.05 (I hope), the AppleTalk module backported to 23.05 and 24.10, and the atalkd problem (on DSA?) narrowed down if not fixed. That’s where time left after paying job and real life will go for a while.

Any particular issues, report as a bug to netatalk. I’m just the messenger.  ;)

**Pull request merged**
